fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/PushReferencedFieldsRequirementThroughDistinctRule.java

```diff
@@ -0,0 +1,76 @@
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.rules;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryFilterPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.temp.ExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRule;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRule.PreOrderRule;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRuleCall;
+import com.apple.foundationdb.record.query.plan.temp.Quantifier;
+import com.apple.foundationdb.record.query.plan.temp.ReferencedFieldsAttribute;
+import com.apple.foundationdb.record.query.plan.temp.ReferencedFieldsAttribute.ReferencedFieldsInfo;
+import com.apple.foundationdb.record.query.plan.temp.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.LogicalDistinctExpression;
+import com.apple.foundationdb.record.query.plan.temp.matchers.BindingMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.PlannerBindings;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatchers;
+import com.google.common.collect.ImmutableSet;
+
+import javax.annotation.Nonnull;
+import java.util.Optional;
+
+import static com.apple.foundationdb.record.query.plan.temp.matchers.ListMatcher.exactly;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.QuantifierMatchers.forEachQuantifierOverRef;
+import static com.apple.foundationdb.record.query.plan.temp.matchers.RelationalExpressionMatchers.logicalDistinctExpression;
+
+/**
+ * A rule that implements a logical filter around a {@link RecordQueryPlan} as a {@link RecordQueryFilterPlan}.
+ */
+@API(API.Status.EXPERIMENTAL)
+@SuppressWarnings("PMD.TooManyStaticImports")
+public class PushReferencedFieldsRequirementThroughDistinctRule extends PlannerRule<LogicalDistinctExpression> implements PreOrderRule {
+    private static final BindingMatcher<ExpressionRef<? extends RelationalExpression>> lowerRefMatcher = ReferenceMatchers.anyRef();
+    private static final BindingMatcher<Quantifier.ForEach> innerQuantifierMatcher = forEachQuantifierOverRef(lowerRefMatcher);
+    private static final BindingMatcher<LogicalDistinctExpression> root =
+            logicalDistinctExpression(exactly(innerQuantifierMatcher));
+
+    public PushReferencedFieldsRequirementThroughDistinctRule() {
+        super(root, ImmutableSet.of(ReferencedFieldsAttribute.REFERENCED_FIELDS));
+    }
+
+    @Override
+    public void onMatch(@Nonnull PlannerRuleCall call) {
+        final PlannerBindings bindings = call.getBindings();
+        final ExpressionRef<? extends RelationalExpression> lowerRef = bindings.get(lowerRefMatcher);
+        final Optional<ReferencedFieldsInfo> requirementOptional = call.getInterestingProperty(ReferencedFieldsAttribute.REFERENCED_FIELDS);
```

Comment:
Is there a way to generalize some of this similar code through all the "Push*" rules, e.g. one rule for all types of requirements on a match?

Reply:
Well, normally they will always do something similar but also everything time in a slightly different context. Ill see what I can do.

Reply:
So naturally each kind of interesting kind of things composes slightly differently with each expression you may want to push it through. It is true that often times you want to do something trivial for the pass-through case that just hands the attribute further down the tree. However, I really don't want to have a generic rule that hands down an attribute blindly if there is no specific method that does something specific to the context. It's better to only do something if it is appropriate and not do it at all if not ok, than to default something here.
Having said that we can only do a implementation reuse that may call out to a common routine that does something like a hand down from within a specific rule. So you would still have to write a rule even though it would be a thin wrapper to a default behavior.

Reply:
I spent some effort to extract the commonalities into common logic that is called from the specific methods. Please take a look.